### DISCUSSION TITLE
Change state in custom component

### DISCUSSION THREAD

**Discussion by aashray16** [Budibase/budibase#11684] [Help] [answered]
Hi, 
I am trying to change a app state in custom component. But finding it difficult to do so.
Also i have a button in that custom component. Can i have onclick options in the settings panel for it. 
Need help for the same.

**Comment by ConorWebb96** [comment]
Hey @aashray16,

I'd recommend checking out how other custom components are done. If its a input field component it should be a `on change` setting that needs added to your schema and, the value of whatever field will need to be configured so its then aviaable as the value within this on change. I've linked a good component below that demos setting the field value up. I'd recomend looking at the schema.json and the component.svelte.

- https://github.com/melohagan/budibase-component-number-counter-input 

For your on click query the schema.json should look something like the below;
```
      {
        "type": "event",
        "key": "btnClicked",
        "label": "On Click"
      },
```

You can check this component https://github.com/mp-strachan/bb-buttonplus/blob/main/src/Component.svelte to see how this has be implemented.

I hope this helps.

**Reply by aashray16** [reply]
Hi @ConorWebb96 ,
My questions are different from one another:
Que1. I want to update this app state in my custom component. Can i anyhow access it in the custom component.svelte file. Is it possible?
![image](https://github.com/Budibase/budibase/assets/88130236/a0780e93-dbaa-4c3e-8060-793922fa80d7)


Que2. While creating a button using button tag, are these onClick setting available for a button in a custom component or i have to manually write script to do the onClick actions.
![image](https://github.com/Budibase/budibase/assets/88130236/7de9ffd4-886a-493e-b78a-9d24f292a95c)

**Reply by ConorWebb96** [reply, marked as answer]
Hey @aashray16,

q1. At present, this functionality is not available. To enable this feature, it would be necessary to expose additional stores within the client library's SDK. I've created an enhancement request for this purpose, this would allow you to modify these actions as described; https://github.com/Budibase/budibase/issues/11746

q2. Yes like i said in the later half of my comment you can attach the button on change event to a button within your custom component.

> For your on click query the schema.json should look something like the below;
>      {
>       "type": "event",
>        "key": "btnClicked",
>        "label": "On Click"
>      },
> You can check this component https://github.com/mp-strachan/bb-buttonplus/blob/main/src/Component.svelte to see how this has be implemented.

<img width="514" alt="Screenshot 2023-09-12 at 09 00 21" src="https://github.com/Budibase/budibase/assets/126772285/89070d80-41ee-4b05-887e-44900b245217">

I hope this helps.

**Reply by aashray16** [reply]
Thanks @ConorWebb96, will try to implement these features. Thanks for the help!!!
